tsl/src/compression/api.c

```diff
@@ -892,3 +896,584 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+									  RelationGetDescr(compressed_chunk_rel));
+}
+
+static bool
+compressed_chunk_column_is_segmentby(PerCompressedColumn per_compressed_col)
+{
+	/* not compressed and not metadata, offset should be -1 for metadata */
+	return !per_compressed_col.is_compressed && per_compressed_col.decompressed_column_offset >= 0;
+}
+
+static void
+decompress_segment_update_current_segment(CompressedSegmentInfo **current_segment,
+										  TupleTableSlot *slot, PerCompressedColumn *per_cols)
+{
+	Datum val;
+	bool is_null;
+	int seg_idx = 0;
+	for (int i = 0; i < slot->tts_tupleDescriptor->natts; i++)
+	{
+		if (!compressed_chunk_column_is_segmentby(per_cols[i]))
+			continue;
+		else
+		{
+			val = slot_getattr(slot, AttrOffsetGetAttrNumber(i), &is_null);
+			if (!segment_info_datum_is_in_group(current_segment[seg_idx++]->segment_info,
+												val,
+												is_null))
+			{
+				/* new segment, need to do per-segment processing */
+				pfree(current_segment[seg_idx - 1]->segment_info); // bacause increased previously
+				SegmentInfo *segment_info =
+					segment_info_new(TupleDescAttr(slot->tts_tupleDescriptor, i));
+				segment_info_update(segment_info, val, is_null);
+				current_segment[seg_idx - 1]->segment_info = segment_info;
+				current_segment[seg_idx - 1]->decompressed_chunk_offset =
+					per_cols[i].decompressed_column_offset;
+			}
+		}
+	}
+}
+
+static bool
+decompress_segment_changed_group(CompressedSegmentInfo **current_segment, TupleTableSlot *slot,
+								 PerCompressedColumn **per_cols)
+{
+	PerCompressedColumn *decompressor_per_compressed_cols = *per_cols;
+	Datum val;
+	bool is_null;
+	bool changed_segment = false;
+	int seg_idx = 0;
+	for (int i = 0; i < slot->tts_tupleDescriptor->natts; i++)
+	{
+		if (!compressed_chunk_column_is_segmentby(decompressor_per_compressed_cols[i]))
+			continue;
+		else
+		{
+			val = slot_getattr(slot, AttrOffsetGetAttrNumber(i), &is_null);
+			if (!segment_info_datum_is_in_group(current_segment[seg_idx++]->segment_info,
+												val,
+												is_null))
+			{
+				changed_segment = true;
+				break;
+			}
+		}
+	}
+	return changed_segment;
+}
+
+static bool
+recompress_remote_chunk(FunctionCallInfo fcinfo, Chunk *chunk)
+{
+	return invoke_compression_func_remotely(fcinfo, chunk);
+}
+
+/*
+ * This is hacky but it doesn't matter. We just want to check for the existence of such an index
+ * on the compressed chunk. For distributed hypertables, returning the index oid would raise issues,
+ * because the Access Node does not see that oid. So we return the oid of the uncompresed chunk
+ * instead, when an index is found.
+ */
+extern Datum
+tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (NULL == uncompressed_chunk)
+		elog(ERROR, "unknown chunk id %d", uncompressed_chunk_id);
```

Comment:
You are calling `ts_chunk_get_by_relid` so that it errors when the chunk does not exist so this code is redundant.

Reply:
right, removed it